core/src/main/scala/cats/data/HashMap.scala

```diff
@@ -0,0 +1,1199 @@
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/* This file is derived from https://github.com/scala/scala/blob/v2.13.8/src/library/scala/collection/immutable/HashMap.scala
+ * Modified by Typelevel for redistribution in Cats.
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ * Scala
+ * Copyright (c) 2002-2022 EPFL
+ * Copyright (c) 2011-2022 Lightbend, Inc.
+ *
+ * Scala includes software developed at
+ * LAMP/EPFL (https://lamp.epfl.ch/) and
+ * Lightbend, Inc. (https://www.lightbend.com/).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats.data
+
+import cats.kernel.compat.scalaVersionSpecific._
+
+import cats.Always
+import cats.CommutativeApplicative
+import cats.Eval
+import cats.Foldable
+import cats.Semigroup
+import cats.Show
+import cats.UnorderedTraverse
+import cats.kernel.CommutativeMonoid
+import cats.kernel.CommutativeSemigroup
+import cats.kernel.Monoid
+import cats.kernel.Eq
+import cats.kernel.Hash
+import cats.kernel.instances.StaticMethods
+import cats.syntax.eq._
+import java.util.Arrays
+
+import HashMap.improve
+
+/**
+ * An immutable hash map using [[cats.kernel.Hash]] for hashing.
+ *
+ * Implemented using the CHAMP encoding.
+ * @see [[https://michael.steindorfer.name/publications/phd-thesis-efficient-immutable-collections.pdf Efficient Immutable Collections]]
+ *
+ * @tparam K the type of the keys contained in this hash map.
+ * @tparam V the type of the values contained in this hash map.
+ * @param hashKey the [[cats.kernel.Hash]] instance used for hashing keys.
+ */
+final class HashMap[K, +V] private[data] (private[data] val rootNode: HashMap.Node[K, V])(implicit val hashKey: Hash[K])
+    extends HashMapCompat[K, V] {
+
+  /**
+    * An iterator for this map that can be used only once.
+    *
+    * @return an iterator that iterates through the key-value pairs of this map.
+    */
+  final def iterator: Iterator[(K, V)] =
+    new HashMap.Iterator(rootNode)
+
```

Comment:
could we add:
```scala
final def toMap: Map[K, V] =
  // add a *wrapper* not a copy of the `HashMap`
```
I think you only need to implement like 4 methods to implement a `Map` and I think you have almost all of them already.

Reply:
I had a crack at this in https://github.com/typelevel/cats/pull/4193/commits/0f663332f848ae4720513bbf57a6c4ea481ae2b0, let me know what you think!